debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/AbstractTransactionCachingLogMinerEventProcessor.java

```diff
@@ -0,0 +1,435 @@
+            // undo entry was read in a prior mining session and the transaction's sequence could not be
+            // resolved. In this case, lets locate the transaction based solely on XIDUSN and XIDSLT.
+            final String transactionPrefix = getTransactionIdPrefix(row.getTransactionId());
+            LOGGER.debug("Undo change refers to a transaction that has no explicit sequence, '{}'", row.getTransactionId());
+            LOGGER.debug("Checking all transactions with prefix '{}'", transactionPrefix);
+            eventKeys = getTransactionKeysWithPrefix(transactionPrefix);
+            if (!eventKeys.isEmpty()) {
+                // Enforce that the keys are always reverse sorted.
+                eventKeys.sort(EventKeySortComparator.INSTANCE.reversed());
+
+                for (String eventKey : eventKeys) {
+                    final LogMinerEvent event = getEventCache().get(eventKey);
+                    if (event != null && event.getRowId().equals(row.getRowId())) {
+                        Loggings.logDebugAndTraceRecord(LOGGER, row, "Undo change on table '{}' applied to transaction '{}'", row.getTableId(), eventKey);
+                        getEventCache().remove(eventKey);
+                        inMemoryPendingTransactionsCache.decrement(row.getTransactionId());
+                        return;
+                    }
+                }
+                Loggings.logWarningAndTraceRecord(LOGGER, row, "Cannot undo change on table '{}' since event with row-id {} was not found.", row.getTableId(),
+                        row.getRowId());
+            }
+            else if (!getConfig().isLobEnabled()) {
+                Loggings.logWarningAndTraceRecord(LOGGER, row, "Cannot undo change on table '{}' since transaction '{}' was not found.", row.getTableId(),
+                        row.getTransactionId());
+            }
+        }
+        else {
+            // Enforce that the keys are always reverse sorted.
+            eventKeys.sort(EventKeySortComparator.INSTANCE.reversed());
+
+            for (String eventKey : eventKeys) {
+                final LogMinerEvent event = getEventCache().get(eventKey);
+                if (event != null && event.getRowId().equals(row.getRowId())) {
+                    LOGGER.debug("Undo applied for event {}.", event);
+                    getEventCache().remove(eventKey);
+                    inMemoryPendingTransactionsCache.decrement(row.getTransactionId());
+                    return;
+                }
+            }
+            Loggings.logWarningAndTraceRecord(LOGGER, row, "Cannot undo change on table '{}' since event with row-id {} was not found.", row.getTableId(),
+                    row.getRowId());
+        }
+    }
+
+    protected List<String> getTransactionKeysWithPrefix(String prefix) {
+        AtomicReference<List<String>> result = new AtomicReference<>();
+        getEventCache().keys(stream -> {
```

Comment:
I am curious: if a user has many in-progress transactions, what overhead in performance does this iteration across all events in the cache cause, particularly if I have multiple large transactions?

Reply:
Logged [DBZ-8011](https://issues.redhat.com/browse/DBZ-8111) to document this so we can review this later and see whether there are any performance concerns with this specific method.